[nnc] Use sleef if its symbols are available

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51252 [nnc] Don't fuse intrinsics on cpu if they are slower
* #51246 [nnc] Don't use sleef where it's slower
* #51190 [nnc] Expose vectorized math functions to jit fuser.
* **#51187 [nnc] Use sleef if its symbols are available**
* #51186 [nnc][trivial] Refactor llvm_jit so the wrapper class doesn't depend on ifdefs
* #51125 [nnc] Refactor generation of intrinsics to reduce the amount of macro-hell

Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

bertmaher committed:
[nnc] Use sleef if its symbols are available

Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]

Krovatkin reviewed (approved):
:shipit:

bertmaher committed:
Update on "[nnc] Use sleef if its symbols are available"


Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]

bertmaher committed:
Fix some compilation issues with gcc5.4, msvc; fix double dtype handling on "[nnc] Use sleef if its symbols are available"


Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]

bertmaher committed:
maybe fix windows on "[nnc] Use sleef if its symbols are available"


Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/llvm_jit.cpp:
```diff
@@ -255,6 +255,10 @@ class TORCH_API PytorchLLVMJITImpl {
     return assertSuccess(LLJ->lookup(Name));
   }
 
+  bool hasSymbol(const std::string& Name) {
```
Ah, that's where it's defined :)

bertmaher committed:
Update on "[nnc] Use sleef if its symbols are available"


Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]

bertmaher committed:
Work around debug-mode llvm really not wanting you to lookup undefined symbols on "[nnc] Use sleef if its symbols are available"


Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]

bertmaher committed:
fix sleef type suffix at the appropriate point in the stack on "[nnc] Use sleef if its symbols are available"


Instead of relying on #ifdefs, we want to use sleef if its symbols are
available.  This diff adds the mechanism to do that check using LLVM's symbol
lookup.

This diff by itself is a no-op, because sleef isn't properly being exposed to
LLVM yet (the `#ifdef __AVX__` checks are always false, because torch/jit isn't
built with `-mavx`).  The next diff will properly expose the symbols, and
perform run time checks.

Differential Revision: [D26096206](https://our.internmc.facebook.com/intern/diff/D26096206/)

[ghstack-poisoned]
